ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/helpers/InListEvaluator.java

```diff
@@ -0,0 +1,283 @@
+      .put(Double.class, Double::valueOf)
+      .put(BigDecimal.class, BigDecimal::new)
+      .build();
+
+  private InListEvaluator() {
+  }
+
+  /**
+   * Looks for {@code value} in {@code values}.
+   *
+   * <p>SQL NULLs never match.
+   *
+   * <p>Conversion from INT to BIGINT is handled.
+   *
+   * <p>Conversion from DECIMAL to DOUBLE is handled.
+   *
+   * <p>Conversion between STRING and other primitive keys is handled.
+   *
+   * <p>Invalid conversions are ignored.
+   *
+   * @param value the value to look for
+   * @param values the values to look in
+   * @return {@code true} if {@code value} is in {@code values}.
+   */
+  public static boolean matches(final Object value, final Object... values) {
+    if (value == null) {
+      // SQL NULL never matches anything:
+      return false;
+    }
+
+    for (final Object v : values) {
+      if (v == null) {
+        // SQL NULL never matches anything:
+        continue;
+      }
+
+      if (isMatch(value, v)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  private static boolean isMatch(final Object requiredValue, final Object possibleMatch) {
+    for (final Handler handler : HANDLERS) {
+      final Optional<Boolean> result = handler.accept(requiredValue, possibleMatch);
+      if (result.isPresent()) {
+        return result.get();
+      }
+    }
+
+    return false;
+  }
+
+  private static Optional<Boolean> nullsNeverMatch(
+      final Object requiredValue,
+      final Object possibleMatch
+  ) {
+    if (requiredValue == null || possibleMatch == null) {
+      // SQL NULL never matches anything:
+      return Optional.of(false);
+    }
+
+    return Optional.empty();
+  }
+
+  private static Optional<Boolean> exactMatch(
+      final Object requiredValue,
+      final Object possibleMatch
+  ) {
+    if (requiredValue.equals(possibleMatch)) {
+      return Optional.of(true);
+    }
+
+    return Optional.empty();
+  }
+
+  private static Object parseString(
+      final String value,
+      final Class<?> requiredType
+  ) {
+    final Function<String, ?> parser = STRING_PARSERS.get(requiredType);
+    if (parser == null) {
+      return NO_MATCH;
+    }
+
+    try {
+      return parser.apply(value);
+    } catch (final NumberFormatException e) {
+      return NO_MATCH;
+    }
+  }
+
+  private static boolean requiredParsedMatches(
+      final String requiredValue,
+      final Object possibleMatch
+  ) {
+    return possibleParsedMatches(possibleMatch, requiredValue);
+  }
+
+  private static boolean possibleParsedMatches(
+      final Object requiredValue,
+      final String possibleMatch
+  ) {
+    final Object parsed = parseString(possibleMatch, requiredValue.getClass());
+    return requiredValue.equals(parsed);
+  }
+
+  private static boolean upCastInt(
+      final Long requiredValue,
+      final Integer possibleMatch
+  ) {
+    return requiredValue.equals(possibleMatch.longValue());
+  }
+
+  private static boolean downCastDecimal(
```

Comment:
Is it not possible to have the reverse as well, BigDecimal is required and Double is matched against?  Is that not the case with the integer/long as well?

It would be good to do some test that cover these casting cases.

Reply:
Good catch. On it...